shotover-proxy/src/transforms/redis/sink_single.rs

```diff
@@ -73,43 +86,230 @@ impl Transform for RedisSinkSingle {
                 Box::pin(tls_stream) as Pin<Box<dyn AsyncStream + Send + Sync>>
             } else {
                 Box::pin(tcp_stream) as Pin<Box<dyn AsyncStream + Send + Sync>>
             };
-            self.outbound = Some(Framed::new(generic_stream, RedisCodec::new()));
+
+            let (outbound_tx, outbound_rx) = Framed::new(generic_stream, RedisCodec::new()).split();
+            let (response_messages_tx, response_messages_rx) = mpsc::unbounded_channel();
+            let (sent_message_type_tx, sent_message_type_rx) = mpsc::unbounded_channel();
+
+            tokio::spawn(server_response_processing_task(
+                outbound_rx,
+                self.pushed_messages_tx.clone(),
+                response_messages_tx,
+                sent_message_type_rx,
+            ));
+            self.connection = Some(Connection {
+                response_messages_rx,
+                sent_message_type_tx,
+                outbound_tx,
+            })
+        }
+
+        let connection = self.connection.as_mut().unwrap();
+
+        for message in &mut message_wrapper.messages {
+            let ty = if let Some(Frame::Redis(RedisFrame::Array(array))) = message.frame() {
+                if let Some(RedisFrame::BulkString(bytes)) = array.first() {
+                    match bytes.to_ascii_uppercase().as_slice() {
+                        b"SUBSCRIBE" | b"PSUBSCRIBE" | b"SSUBSCRIBE" => MessageType::Subscribe,
+                        b"UNSUBSCRIBE" | b"PUNSUBSCRIBE" | b"SUNSUBSCRIBE" => {
+                            MessageType::Unsubscribe
+                        }
+                        b"RESET" => MessageType::Reset,
+                        _ => MessageType::Other,
+                    }
+                } else {
+                    MessageType::Other
+                }
+            } else {
+                MessageType::Other
+            };
+            connection.sent_message_type_tx.send(ty).unwrap();
         }
 
-        // self.outbound is gauranteed to be Some by the previous block
-        let outbound_framed_codec = self.outbound.as_mut().unwrap();
-        outbound_framed_codec
+        let messages_len = message_wrapper.messages.len();
+        connection
+            .outbound_tx
             .send(message_wrapper.messages)
             .await
             .ok();
 
-        match outbound_framed_codec.next().fuse().await {
-            Some(mut a) => {
-                if let Ok(messages) = &mut a {
-                    for message in messages {
-                        if let Some(Frame::Redis(RedisFrame::Error(_))) = message.frame() {
-                            self.failed_requests.increment(1);
+        let mut result = Vec::with_capacity(messages_len);
+        while result.len() < messages_len {
+            let mut message = connection.response_messages_rx.recv().await.unwrap();
+            if let Some(Frame::Redis(RedisFrame::Error(_))) = message.frame() {
```

Comment:
Not your code, but there's an easy perf win here if we can avoid calling frame on every message to check for errors.

Reply:
Oh yeah, could be!
You can raise an issue if you want, definitely out of scope of this PR.
And I'm not sure how expensive it is to parse redis responses so we should actually do some perf measurements before blindly making this more complicated.

Reply:
Just commenting it out gives us 20%~ gain on the benchmarks, I'll raise the issue.

Reply:
woah! :eyes:

Reply:
I cant reproduce those results on my machine.
I am getting up to 20% noise variation with the `frame()` and I'm not seeing any improvements with the `frame()` removed.

Reply:
True, though reasonable to assume that not fully parsing every frame will give us a speedup, so I'll leave the issue open for investigation.